src/plugins/intel_cpu/src/memory_state.cpp

```diff
@@ -1,18 +1,131 @@
+                                                     MemoryDescPtr external_desc,
+                                                     MemoryCPtr init_val) :
+    IVariableState{name}, m_external_desc{external_desc} {
+    reset_prime_mem(mem_build());
+    reset_second_mem(mem_build());
+    m_internal_desc = prime_mem()->getDescPtr();
+    auto&& shape = m_internal_desc->getShape();
+    //TODO what if by some reason we already have internal static state while the node is dynamic, is it even possible?
+
+    if (shape.isStatic()) {
+        if (init_val) {
+            prime_mem()->load(*init_val);
+        } else {
+            prime_mem()->nullify();
+        }
+    } else {
+        //in the case of the original desc has dynamic shape we create an empty tensor
+        auto new_desc = to_static(m_internal_desc);
+        prime_mem()->redefineDesc(new_desc);
+    }
+}
+
+void VariableStateDoubleBuffer::set_state(const ov::SoPtr<ov::ITensor>& state) {
+    m_state = state; // simply to extend the lifetime
+    auto state_desc = MemoryDescUtils::generateCpuBlockedMemoryDesc(m_state);
+
+    const auto& shape = state_desc->getShape();
+
+    if (prime_mem()->getShape() != shape) {
+        auto new_desc = m_internal_desc->cloneWithNewDims(shape.getStaticDims());
+        prime_mem()->redefineDesc(new_desc);
+    }
+
+    auto src = m_state->data();
+
+    Memory mem(get_engine(), state_desc, src);
+    prime_mem()->load(mem);
+}
+
+const dnnl::engine& VariableStateDoubleBuffer::get_engine() const {
+    static const dnnl::engine eng(dnnl::engine::kind::cpu, 0);
+    return eng;
+}
+
+ov::SoPtr<ov::ITensor> VariableStateDoubleBuffer::get_state() const {
+    //TODO , in general case must be synchronized
+    const auto& current_dims = prime_mem()->getStaticDims();
+    auto current_ext_desc = m_external_desc->cloneWithNewDims(current_dims);
+    auto current_internal_desc = prime_mem()->getDescPtr();
+
+    if (current_ext_desc->isCompatible(*current_internal_desc)) {
+        return std::make_shared<Tensor>(prime_mem());
+    }
+
+    //test precision
+    {
+        auto internal_prc = current_internal_desc->getPrecision();
+        auto tmp_desc = current_ext_desc->cloneWithNewPrecision(internal_prc);
+        if (tmp_desc->isCompatible(*current_internal_desc)) {
+            auto mem = std::make_shared<Memory>(get_engine(), current_ext_desc);
+            size_t elements_to_convert = prime_mem()->getDescWithType<BlockedMemoryDesc>()->getPaddedElementsCount();
+            auto external_prc = current_ext_desc->getPrecision();
+
+            cpu_convert(prime_mem()->getData(), mem->getData(), internal_prc, external_prc, elements_to_convert);
+            return std::make_shared<Tensor>(mem);
+        }
+    }
+
+    //reorder
+    auto mem = std::make_shared<Memory>(get_engine(), current_ext_desc);
+    mem->load(*(prime_mem()));
+    return std::make_shared<Tensor>(mem);
+}
```

Comment:
Do you have a unit test that triggers this precision conversion?

I checked that the existing PR implementation will always have the same external_desc and internal_desc, as both them comes from getOriginalOutputPrecisionAtPort(0). Is it what we expect? 
I have a [comment](https://github.com/openvinotoolkit/openvino/pull/20404#discussion_r1395091262) and try to figure out if there any way to let a user declare his wanted precision to a state. It is useful we use half-precision (FP16) kvcaches for platforms of avx2/avx512.

Reply:
We don't. I think we may do it in a separate PR since this one is necessary to unblock some other activities.